### ISSUE TITLE
Implement setting of freedesktop.org EWHM _NET properties

### ISSUE THREAD

**Issue by tomterl** [jwilm/alacritty#348]
There are certain properties specified by freedesktop.org that desktop environments rely on, but that aren't set by the window-manager -- so the client should set them itself.

Here is the list: https://specifications.freedesktop.org/wm-spec/1.3/ar01s05.html

It would be easier/possible to configure enhanced window-manager behaviour regarding alacritty if these properties were present.

Not all of the properties are necesseary, but all marked with 'the client SHOULD/COULD set' should be considered (_NET_WM_NAME, _NET_WM_PID, NET_WM_WINDOW_TYPE, ...)

**Comment by jonhoo** [comment]
The fact that this is missing also means that it's hard to write automatic window placement rules for applications running inside alacritty. For example, I want `mutt` to always be run on my sixth virtual desktop, which I do in `xmonad` by matching on `WM_NAME`. Since alacritty does not set `WM_NAME`, this information is not communicated to the window manager, and thus the window is not placed where I want it to.

**Comment by jwilm** [comment]
@jonhoo as a workaround, can you use the window title until that's added?

**Comment by jonhoo** [comment]
Ah, so interestingly `WM_NAME` *does* contain "mutt" when I run `mutt` directly from my shell within alacritty. However, when I run `mutt` inside a `tmux` session, `WM_NAME` is *not* set by alacritty (URxvt handles this perfectly fine). I don't know if that's a related problem?

Here's a completely random, and probably incorrect guess:
Maybe mutt sets the `WM_NAME` on its parent directly somehow. When run without tmux, this sets the name of the alacritty window. When `tmux` is running, however, it sets the name of the `tmux` "window" instead. URxvt propagates this up to the `WM_NAME` of the "top-level" window, but alacritty does not (which is what #348 is all about).

**Comment by jonhoo** [comment]
@jwilm this is biting me in more and more cases lately where I want to assign particular window placement rules to alacritty instances that wrap different programs. Any idea how hard this would be to implement?

**Comment by jwilm** [comment]
@jonhoo I [added](https://github.com/jwilm/alacritty/blob/master/src/window.rs#L330-L339) `_NET_WM_PID`, and it shouldn't be too hard to add the others now that one exists.

This is kinda low priority for me at the moment (working on some correctness items and getting glutin updated), but I would be happy to look at a PR if it came in.

**Comment by jwilm** [comment]
Oh, just took a second look at your comments. It seems like you must have some tmux setting (or combination with urxvt) which is enabling that behavior. Would you be ok sharing those config files?

**Comment by jwilm** [comment]
@tomterl is there a specific property you're looking for that's not available?

**Comment by jonhoo** [comment]
Ahh, I did some more digging, and it turns out the issue is *actually* that I'm launching alacritty with `-e <command>`. When launched like that, it just sets its name to "Alacritty", not the name of the executing program. If I launch mutt (or any other program) inside tmux inside alacritty normally (i.e., from my shell), the `WM_NAME` changes correctly. So it sounds like `-e` should just be modified to also update `WM_NAME`?

**Comment by jwilm** [comment]
@jonhoo can you try adding something like this to your tmux config?

```
set -g set-titles on
set -g set-titles-string "#T"
```

`set-titles` enables tmux setting the window title, and `set-titles-string` controls how that looks. The example I gave uses the pane title.

**Comment by jonhoo** [comment]
I already have both of those, and tmux correctly sets the title when I run it from inside alacritty. It is specifically when I run tmux using `alacritty -e` that the title isn't set.

Also, as a somewhat separate observation, it looks like alacritty *first* launches, and *then* sets its window title. I just tried adding a rule to my WM to move any window whose `WM_NAME` and `WM_CLASS` are empty *at launch* to a different workspace, and every alacritty window is moved there. Is there any delay in alacritty between when the window is created and the `WM_*` attributes are set?

**Comment by jwilm** [comment]
> I already have both of those, and tmux correctly sets the title when I run it from inside alacritty. It is specifically when I run tmux using alacritty -e that the title isn't set.
 
Got it; thanks for the clarification.


> Is there any delay in alacritty between when the window is created and the WM_* attributes are set?

Not intentionally. We use Glutin's `WindowBuilder` to provide the title at creation, but it's possible Glutin applies it after. Would need to dig into that code to get a full understanding.

I'm still confused about why it would just say Alacritty given your tmux config. I see the title being set when running with `-e tmux`.

**Comment by jonhoo** [comment]
```console
$ alacritty -e tmux new-session -A -s mail /bin/bash -ic mutt
[window launches]
$ xprop
WM_LOCALE_NAME(STRING) = "C"
WM_CLIENT_MACHINE(STRING) = "x1c3"
_NET_WM_PID(CARDINAL) = 26207
_NET_WM_NAME(UTF8_STRING) = "x1c3"
WM_NAME(STRING) = "x1c3"
WM_NORMAL_HINTS(WM_SIZE_HINTS):
                program specified size: 1024 by 768
WM_CLASS(STRING) = "Alacritty", "Alacritty"
_NET_WM_DESKTOP(CARDINAL) = 1
WM_STATE(WM_STATE):
                window state: Normal
                icon window: 0x0
WM_PROTOCOLS(ATOM): protocols  WM_DELETE_WINDOW
```

**Comment by jonhoo** [comment]
It looks like the delayed title/class setting is indeed a glutin bug. I'll file one shortly and link it here. `WM_CLASS` and `WM_NAME` not being set with `-e` still seems like an `alacritty` bug though.

**Comment by jwilm** [comment]
Does X actually support providing a title before window creation? I'm not super familiar with how that works.

**Comment by jonhoo** [comment]
Filed as https://github.com/tomaka/glutin/issues/879.
Yes, I believe you should be able to set `WM_NAME` and `WM_CLASS` before the window is mapped. Certainly other programs do this.

**Comment by jwilm** [comment]
@jonhoo "mapping" does seem to be the critical point. It seems that you can't set the title before creating the window, but it doesn't matter until calling `XMapWindow` when the WM would take some action. Incidentally, I don't see winit calling `XMapWindow`.

**Comment by jonhoo** [comment]
Hmm, curious. It must be mapping it somewhere, otherwise the window wouldn't be visible... :/ I've updated the bug above to say "mapping" instead of "creating".

**Comment by jonhoo** [comment]
There's certainly code there for setting the name and class in `winit::Window::new`: https://github.com/tomaka/winit/blob/master/src/platform/linux/x11/window.rs#L469. But `MapRaised` seems to be called first: https://github.com/tomaka/winit/blob/master/src/platform/linux/x11/window.rs#L409

**Comment by jwilm** [comment]
@jonhoo nice work tracking that down :D

**Comment by jonhoo** [comment]
Heh, thanks. Phew, filed a PR over at https://github.com/tomaka/winit/pull/168 which fixes the issue afaict. I assume that will also fix the issue for alacritty (though not the `-e` problem -- that seems separate).

**Comment by jwilm** [comment]
That's great! Indeed, that should fix the issue here too. I'm currently working on getting up-to-date with glutin/winit and should have something this weekend. Will make sure your patch is included.

**Comment by jonhoo** [comment]
Can you verify that you also see "Alacritty" as the `WM_CLASS`, and your hostname as `WM_NAME` when running the command in https://github.com/jwilm/alacritty/issues/348#issuecomment-299307408 ?

**Comment by jwilm** [comment]
Here's what I get. 
```
WM_LOCALE_NAME(STRING) = "C"
WM_CLIENT_MACHINE(STRING) = "jwilm-desk"
_NET_WM_PID(CARDINAL) = 10611
_NET_WM_NAME(UTF8_STRING) = "jwilm-desk"
WM_NAME(STRING) = "jwilm-desk"
WM_NORMAL_HINTS(WM_SIZE_HINTS):
                program specified size: 1024 by 768
WM_CLASS(STRING) = "Alacritty", "Alacritty"
WM_STATE(WM_STATE):
                window state: Normal
                icon window: 0x0
WM_PROTOCOLS(ATOM): protocols  WM_DELETE_WINDOW
```

**Comment by jonhoo** [comment]
Ah, so basically the same thing that I do. Interestingly, it seems like `urxvt` does *mostly* the same thing:
```
_NET_WM_DESKTOP(CARDINAL) = 2
WM_STATE(WM_STATE):
                window state: Normal
                icon window: 0x0
_NET_WM_PID(CARDINAL) = 1802
WM_PROTOCOLS(ATOM): protocols  WM_DELETE_WINDOW, _NET_WM_PING
WM_LOCALE_NAME(STRING) = "en_US.UTF-8"
WM_CLASS(STRING) = "urxvt", "URxvt"
WM_HINTS(WM_HINTS):
                Client accepts input or input focus: True
                Initial state is Normal State.
                window id # of group leader: 0x500000a
WM_NORMAL_HINTS(WM_SIZE_HINTS):
                program specified minimum size: 12 by 22
                program specified resize increment: 8 by 18
                program specified base size: 4 by 4
                window gravity: NorthWest
WM_CLIENT_MACHINE(STRING) = "x1c3"
WM_COMMAND(STRING) = { "urxvt", "-e", "tmux", "new-session", "-A", "-s", "mail", "/bin/bash", "-ic", "mutt" }
_NET_WM_ICON_NAME(UTF8_STRING) = "tmux"
WM_ICON_NAME(STRING) = "tmux"
_NET_WM_NAME(UTF8_STRING) = "x1c3"
WM_NAME(STRING) = "x1c3"
```
So the current behavior of alacritty is probably correct. Note the `WM_COMMAND` though, which urxvt exposes, and alacritty does not. That would be really useful for doing window manager matching on `-e` windows. Alternatively I suppose once my patch lands I can use `-t` to set a name manually...

**Comment by jwilm** [comment]
@jonhoo Running with `-t` will also set the WM_CLASS as it turns out. Setting WM_COMMAND does seem like a good idea.

**Comment by jonhoo** [comment]
@jwilm yeah, it seems like winit sets both or neither. Though without my patch, `-t` is useless from a window manager perspective, since the given name/class isn't set as far as the window manager knows.

**Comment by jwilm** [comment]
I don't quite understand that last statement. It seems like your patch makes the name/class available to the WM.

**Comment by jonhoo** [comment]
Yes, hence *without* my patch `-t` is useless for the WM :p

**Comment by jwilm** [comment]
Oh, I read _without_ as _with_ several time... never mind me :P

**Comment by jonhoo** [comment]
@jwilm the PR has now landed in winit/glutin, so we should be good for the next release!

**Comment by jwilm** [comment]
@jonhoo that's great! I'm impressed you were able to get that pushed through so quickly.

I just submitted another PR to winit which should allow us to get off the fork. I'll be working on getting it integrated with Alacritty this week :tada:.

**Comment by maximbaz** [comment]
The WM_CLASS is being set now on startup thanks to #671 🙂

**Comment by tomterl** [comment]
Sorry for the long delay -- I just now fixed up my rust setup, so that alacritty compiles again; 

My issue is fixed, xmonand interaction with alacritty works fine. Thanks al lot! alacritty is now set as my standard terminal...

**Comment by irb** [comment]
Would this be the correct place to add that setting WM_COMMAND might also be useful? 

I'm looking to replace urxvt with alacritty, and I'd like to rely on KDE Plasma's session management to restore about 20 terminals across 12 desktops. KDE relies on WM_COMMAND to restore its sessions and since alacritty doesn't set that, I end up having to start them manually. I can script it to some extent, but I'd prefer not to have to. FWIW, I took a look at kitty and it doesn't do this either (and the devs there kindly aren't interested in implementing it). 

Anyway, thanks for an awesome, modern terminal.
